### PR TITLE
Feature/curve lp actions

### PR DESCRIPTION
**Pull request by kjaredic**

### CODE REVIEW

File: contracts/actions/curve/CurveSwap.sol [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import "../ActionBase.sol";
+
+contract CurveSwap is ActionBase, CurveHelper {
+    using TokenUtils for address;
+
+    struct Params {
+        address sender;
+        address receiver;
+        address pool;
+        address from;
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/CurveSwap.sol:17]: tokenA and tokenB

File: contracts/actions/curve/helpers/CurveHelper.sol [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.7.6;
+pragma experimental ABIEncoderV2;
+
+import "../../../interfaces/curve/IAddressProvider.sol";
+import "../../../interfaces/curve/ISwaps.sol";
+
+contract CurveHelper {
+    int128 constant MAX_COINS = 8;
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/helpers/CurveHelper.sol:10]: Check if needed

Comment (yippee-ki-yay) [reply contracts/actions/curve/helpers/CurveHelper.sol:10]: CALC_INPUT_SIZE as well

File: contracts/actions/curve/CurveDeposit.sol [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import "../../utils/SafeMath.sol";
+import "../ActionBase.sol";
+
+contract CurveDeposit is ActionBase, CurveHelper {
+    using TokenUtils for address;
+    using SafeMath for uint256;
+
+    function _sig(uint256 _nCoins, bool _useUnderlying) internal pure returns(bytes4) {
+        if (!_useUnderlying) {
+            if (_nCoins == 2) return bytes4(0x0b4c7e4d);
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/CurveDeposit.sol:17]: keccak(string) to see the funcSig in readable form

File: contracts/actions/curve/CurveDeposit.sol [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import "./helpers/CurveHelper.sol";
+import "../../utils/TokenUtils.sol";
+import "../../utils/SafeMath.sol";
+import "../ActionBase.sol";
+
+contract CurveDeposit is ActionBase, CurveHelper {
+    using TokenUtils for address;
+    using SafeMath for uint256;
+
+    function _sig(uint256 _nCoins, bool _useUnderlying) internal pure returns(bytes4) {
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/CurveDeposit.sol:15]: Maybe move all this to front, as a public view they call before hand

File: contracts/actions/curve/CurveDeposit.sol [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            mstore(add(payload, 0x20), sig)    // store the signature after dynamic array length field (&callData + 0x20)
+
+            let offset := 0x20  // offset of the first element in '_amounts'
+            for { let i := 0 } lt(i, mload(_amounts)) { i := add(i, 1) } {
+                mstore(add(payload, add(0x4, offset)), mload(add(_amounts, offset)))    // payload offset needs to account for 0x4 bytes of selector
+                offset := add(offset, 0x20) // offset for next copy
+            }
+
+            mstore(add(payload, add(0x4, offset)), _minMintAmount)
+            offset := add(offset, 0x20) // offset for final conditional copy
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/CurveDeposit.sol:137]: can move down

File: contracts/actions/curve/CurveDeposit.sol [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                msgValue = _params.amounts[i];
+                continue;
+            }
+            _params.tokens[i].pullTokensIfNeeded(_params.sender, _params.amounts[i]);
+            _params.tokens[i].approveToken(_params.depositTarget, _params.amounts[i]);
+        }
+
+        bytes memory payload = _constructPayload(_params.amounts, _params.minMintAmount, _params.useUnderlying);
+        (bool success, ) = _params.depositTarget.call{ value: msgValue }(payload);
+        require(success);
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/CurveDeposit.sol:102]: err msg

File: contracts/actions/curve/CurveWithdraw.sol [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        _params.lpToken.pullTokensIfNeeded(_params.sender, _params.burnAmount);
+        _params.lpToken.approveToken(_params.pool, _params.burnAmount);
+        
+        uint256[] memory balances = new uint256[](_params.tokens.length);
+        for (uint256 i = 0; i < _params.tokens.length; i++) {
+            balances[i] = _params.tokens[i].getBalance(address(this));
+        }
+
+        bytes memory payload = _constructPayload(_params.minAmounts, _params.burnAmount, _params.useUnderlying);
+        (bool success, ) = _params.pool.call(payload);
```
Comment (yippee-ki-yay) [inline contracts/actions/curve/CurveWithdraw.sol:95]: Is there a way to validate if curve pool?

**Review by yippee-ki-yay — COMMENTED** [review]
Need some additional changes and to see if we can avoid assembly

File: contracts/actions/curve/CurveSwap.sol (hunk 1 above)
Comment (kjaredic) [reply contracts/actions/curve/CurveSwap.sol]: done

File: contracts/actions/curve/helpers/CurveHelper.sol (hunk 2 above)
Comment (kjaredic) [reply contracts/actions/curve/helpers/CurveHelper.sol:10]: not needed

File: contracts/actions/curve/CurveDeposit.sol (hunk 3 above)
Comment (kjaredic) [reply contracts/actions/curve/CurveDeposit.sol]: done, moved to CurveView

File: contracts/actions/curve/CurveDeposit.sol (hunk 4 above)
Comment (kjaredic) [reply contracts/actions/curve/CurveDeposit.sol]: done

File: contracts/actions/curve/CurveWithdraw.sol (hunk 7 above)
Comment (kjaredic) [reply contracts/actions/curve/CurveWithdraw.sol]: keep mapping of pool addresses in View contract and add it to refistr?

File: contracts/actions/curve/CurveDeposit.sol (hunk 6 above)
Comment (kjaredic) [reply contracts/actions/curve/CurveDeposit.sol]: aye

File: contracts/actions/curve/CurveDeposit.sol (hunk 5 above)
Comment (kjaredic) [reply contracts/actions/curve/CurveDeposit.sol]: yes

**Pull request closed without merging**